### ISSUE TITLE
Example code for usage demonstration - request for some use cases

### ISSUE THREAD

**Issue by ajithc16** [enioka/jqm#237]
Thank you JQM developers, I could follow the documentation and set JQM node with oracle back-end. I checked in oracle side and found tables are created fine. Documentation was helpful up to that stage. I could not follow remaining part.

1. I have a java swing desktop application, which collects some meta information and associated files (large size) - this swing gui will be invoked using JNLP and run in client desktop. 
2. I have another java jar - this need to run in server, take arguments, parses all files and update oracle database 

How do I submit the job from my swing application, so that job get queued and executed in server side.
JQM is already setup in server, I wish to use the a particular queue in JQM for batch job job handling. Where do/how do I specify the host/node/queue details? Or is it like I can directly update the oracle table rows?

I tried to read and understand the flow from available documentation, I am lost. May be I missed some documentation/steps?

If you can share some use cases, it would be well appreciated. Ideally we should have a front-end application (Swing/web - asps/jsp) and back end code. 

I also wish to offload some batch (long sql queries/report generation) from aspx web portal. So an example involving job submission from aspx code also will be appreciated. 

Another use case would be launching a shell script in the server using queue in JQM

Adding these and some more use cases to our documentation can help beginners like me to get started fast with JQM.

**Comment by marcanpilami** [comment]
Hi,

for putting new job execution requests in a queue, please look at the client API (http://jqm.readthedocs.io/en/jqm-all-1.3.7/client/index.html). From a swing heavy client, I would heavily suggets using the REST API variant (http://jqm.readthedocs.io/en/jqm-all-1.3.7/client/webservice.html) which has a ready to use java library.

For submitting jobs from an aspx (so I believe .NET) application, please use the same API. However, you will have to call the services yourself, as there is no ready to use library for .NET. The services are fully described in the documentation and quite trivial, so this should not be an issue.

For creating queues and job definitions, please use the integrated admin GUI (http://jqm.readthedocs.io/en/jqm-all-1.3.7/admin/gui.html).

Does this answer your questions?

**Comment by ajithc16** [comment]
I have been going through the docs for some time. I still could not understand many things (may be due to lack of my knowledge of JPA etc..)

I am still (re)reading the docs. 
5.3.2.3. Example in docs says **int i = JqmClientFactory.getClient().enqueue("superbatchjob");**
But there is not such API exists, it could be one of the following
1. enqueue(JobRequest jr) 
2. enqueue(String string, String string1)

Apart from the unit test codes, readable and ready to run code can help. I will read the RESET api again. 

Thank you

**Comment by ajithc16** [comment]
Due to large number of dependencies for WS API, I started digging into JPA client API. After lot of struggle with dependencies, I could create a client jar successfully.

I am trying to run the simple enqueue code, like this

        JobRequest jr = new JobRequest("epinet.batchloader.BatchLoader2D", "user", "email");
        int i = JqmClientFactory.getClient().enqueue(jr);

While running, I am java -jar -Dlog4j.debug seisqel.jar, I am getting following error. 
**3402 [main] ERROR com.enioka.jqm.api.HibernateClient  - Job definition named epinet.batchloader.BatchLoader2D does not exist
Exception in thread "main" com.enioka.jqm.api.JqmInvalidRequestException: no job definition named epinet.batchloader.BatchLoader2D
	at com.enioka.jqm.api.HibernateClient.enqueue(HibernateClient.java:267)
	at epinet.seisqel.JQMTest.main**

I think I am able to use JPA at client side successfully. I think I need to define a job xml and import into server end now? Does this mean for every job we are planning to execute, JAR and job defs should be present in the server? 

My server job takes lot of paramters through a xml file. Now my question is how do I pass this xml file to job? 

Here is my work flow
1. User starts the swing app in one workstation
2. He selects files for processing
3. List of files and parameters need to be submitted to server
4. Server will start my job with a xml parameter file (any other option to pass parameters?)
6. Logs will be preserved for later use

Can you give some leads to point 3 and 4?
Thank you

**Comment by marcanpilami** [comment]
Hi,

I'm sorry to hear you had a lot of trouble with dependencies. Could you please elaborate? It is critical that the libs are easy to use, so it would help a lot other users. In the meantime, usually the WS or the JPA clients only one has one top-level dependency. If you use Maven (but it also work with Gradle & other modern build systems with a different syntax), using the lib is simply a matter of putting inside your pom
```
<dependency>
        <groupId>com.enioka.jqm</groupId>
        <artifactId>jqm-api-client-jersey</artifactId>
        <version>1.3.7</version>
</dependency>
```
And I would *strongly* advise against using the JPA API in a swing heavy/fat client - you are making a dangerous shortcut in your architecture by connecting the client to the database. It may very well come back to bite you (for example, you'll need no firewalls between clients and database... not a good security practice. Or : you'll need database credentials in the client ! Which is even worst). The JPA API is first and foremost for server-side applications. Also, I should stress that when taking transitive dependencies into account, there are a lot more libraires needed for the JPA version than the WS one.


As for later error, as it states there is no job definition defined in JQM named like this. Job definitions are the "job instance templates" - they define what should be run (where the jar is, the name of the class, the class loading options...). So yes you should define a JobDef with the admin GUI. For more on the basic concepts, see http://jqm.readthedocs.io/en/jqm-all-1.3.7/archi.html

For passing parameters : default parameters are defined in the jobdef. On can add others (and overload defaults) at runtime in your JobRequest.

Also, please note that the JobRequest fluent API is easier to use : 
```
JobRequest.create("myjobdefname", "myuser").addParameter("parameter_key", "value").addParameter(...).submit();
```

**Comment by marcanpilami** [comment]
For the team: this ticket is classified as doc. We should add a "my first client app" tutorial in the doc.

**Comment by ajithc16** [comment]
Glad to get your support, your links helped me so far.

I am developing the swing app and batch job, both in Netbeans. So I had to
add required jars one by one to project library list. I don't use maven.
Each run reports some No class found exception, I add that jar from the JQM
lib folder to NB project dependency list.

I tried adding jersey lib to NB, it added lot of jars to the list. So I
stopped that route and tried JPA. As per your suggestion, I will try WS API
tomorrow.

While adding JQM WS in netbeans services list, it is asking for a WSDL or
WADL file for the WS. How can I define that? When I gave the WS URI in
netbeans as http://myserer:8832/ws/client (sorry I don't remember the exact
URL), it prompted a login dialog with JQM as title and asked for user name
password. But login did not succeed, error was netbeans could not recognise
if the WS is WSDL or WADL.

Idea of "My first JQM app" is very good. We can have examples for
1. Web app JPA
2. Swing app WS using jersey
3. WS for non java like python or aspx etc..

I am happy to reach till this point, in fact very excited and hopeful of
making it work.

Ajith

**Comment by marcanpilami** [comment]
Hi,

I'm afraid you are taking the long route for using the WS client. First, there is no WSDL. WSDL is for SOAP web services, and the JQM API is a REST web service. And since you are using Java, you do not have to deal with the web services yourself - JQM provides a client library for Java. The one I refer to in a previous comment, named jqm-api-client-jersey. Please read the actual doc - we do not have a tutorial, but we do have examples. 

Also, NetBeans has excellent Maven integration. I suggest you read one of the many tutorials floating around. You may not want to use it, you may want to use something else, but keep in mind that NetBeans is a very rich product integrated with most build systems.

Finally, putting your batch jobs jar files inside the "lib" directory is weird. It works, but this directory only exists for drivers. You loose a lot by using it (CL isolation, change detection, Maven integration...) Your jobs are supposed to be somewhere of your choosing - and this place must be referenced in the job definition, so that the engine knows where to look for it.

**Comment by ajithc16** [comment]
I am using Netbeans default ant build, so not switching over to maven now.
After including required jar files in dependency list, I could compile and run the example code using REST jersey api.

```
package epinet.seisqel;
import com.enioka.jqm.api.JobInstance;
import com.enioka.jqm.api.JobRequest;
import com.enioka.jqm.api.JqmClientFactory;
import com.enioka.jqm.api.Queue;
import com.enioka.jqm.api.State;
import java.util.List;
import org.apache.log4j.Logger;

public class JQMTest {
    private static Logger jqmlogger = Logger.getLogger(JQMTest.class);
    
    public static void main(String[] args) {
        jqmlogger.info("Testing JQM enqueue");
        
        List<Queue> jq = JqmClientFactory.getClient().getQueues();
        System.out.println(jq.size());
        System.out.println(jq.get(0).getName());

        // running jobs
        List<JobInstance> ji = JqmClientFactory.getClient().getJobs();
        System.out.println(ji.size());
        
        // Enqueue a job
        int i = JqmClientFactory.getClient().enqueue("DemoFibo1", "e94342");        
        
        // Get its status
        State s = JqmClientFactory.getClient().getJob(i).getState();

        // If still waiting, cancel it
        if (s.equals(State.RUNNING)) {
            JqmClientFactory.getClient().cancelJob(i);
        }
    }
}
```
I had to include following dependencies (log4j, JAX-RS-2.0 and Jersey 2.5.1)
After build, additional 29 jar files from above libraries were in lib directory. Code run fine and printed results for first call to getQueues(), subsequent call to getJobs() or enqueue() failed
I have 4 queues defined, first is DEFAULT queue.

Here is the log,

> 0    [main] INFO  epinet.seisqel.JQMTest  - Testing JQM enqueue
> 22   [main] INFO  com.enioka.jqm.api.JerseyClient  - Loading jqm properties file
> 68   [main] INFO  com.enioka.jqm.api.JerseyClient  - A login/password pair was specified and will be used
> 93   [main] DEBUG com.enioka.jqm.api.JerseyClient  - The following root service URL will be used: http://psschn01:8832/ws/client
> 4
> DEFAULT
> Exception in thread "main" com.enioka.jqm.api.JqmClientException: an internal JQM client exception occured
> 	at com.enioka.jqm.api.JerseyClient.getJobs(JerseyClient.java:482)
> 	at epinet.seisqel.JQMTest.main(JQMTest.java:33)
> Caused by: javax.ws.rs.NotAuthorizedException: HTTP 401 Unauthorized
> 	at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:911)
> 	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:822)
> 	at org.glassfish.jersey.client.JerseyInvocation.access$600(JerseyInvocation.java:90)
> 	at org.glassfish.jersey.client.JerseyInvocation$3.call(JerseyInvocation.java:693)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
> 	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:422)
> 	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:689)
> 	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:405)
> 	at org.glassfish.jersey.client.JerseyInvocation$Builder.get(JerseyInvocation.java:301)
> 	at com.enioka.jqm.api.JerseyClient.getJobs(JerseyClient.java:476)
> 	... 1 more

I have jqm and log4j properties setup correctly. Infact first call to getQueues() worked.
What could have caused the javax.ws.rs.NotAuthorizedException: HTTP 401 Unauthorized?

**Comment by ajithc16** [comment]
I created a new maven project and tried with the code. Dependency declaration was breeze. 
But while running after successfully completing getQueues(), I am getting the same error for subsequent call.

```

> Exception in thread "main" com.enioka.jqm.api.JqmClientException: an internal JQM client exception occured
> 	at com.enioka.jqm.api.JerseyClient.enqueue(JerseyClient.java:263)
> 	at epinet.seisqel2.JQMTest.main(JQMTest.java:33)
> Caused by: javax.ws.rs.NotAuthorizedException: HTTP 401 Unauthorized
> 	at org.glassfish.jersey.client.JerseyInvocation.convertToException(JerseyInvocation.java:911)
> 	at org.glassfish.jersey.client.JerseyInvocation.translate(JerseyInvocation.java:770)
> 	at org.glassfish.jersey.client.JerseyInvocation.access$500(JerseyInvocation.java:90)
> 	at org.glassfish.jersey.client.JerseyInvocation$2.call(JerseyInvocation.java:671)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:315)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:297)
> 	at org.glassfish.jersey.internal.Errors.process(Errors.java:228)
> 	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:423)
> 	at org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:667)
> 	at org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:423)
> 	at org.glassfish.jersey.client.JerseyInvocation$Builder.post(JerseyInvocation.java:326)
> 	at com.enioka.jqm.api.JerseyClient.enqueue(JerseyClient.java:255)
> 	... 1 more

```

**Comment by ajithc16** [comment]
With referece to jersey api,

- Any call with out argument works.
- Any call with arguments fails with HTTP 401 Unauthorized.

I could not work around this.

**Comment by ajithc16** [comment]
I am through enqueue() finally. Doc subsection [6.11.3.1.2. Authentication](http://jqm.readthedocs.io/en/jqm-all-1.3.7/admin/security.html) gave the clue.

I found that there are 2 users in RUSER table in JQM database, root and another one, but both were not with WS-API role (31).
So manually inserted a new user jqmwsc with required role 
	java -jar ./jqm.jar -U jqmwsc,jqmwsp,"client power user"

Updated jqm.properties files and test job enqueue() run fine!!
